internal/services/netapp/netapp_volume_helper.go

```diff
@@ -0,0 +1,978 @@
+	}
+}
+
+func netappVolumeReplicationMirrorStateRefreshFunc(ctx context.Context, client *volumesreplication.VolumesReplicationClient, id volumesreplication.VolumeId, desiredState string) pluginsdk.StateRefreshFunc {
+	validStates := []string{"mirrored", "broken", "uninitialized"}
+
+	return func() (interface{}, string, error) {
+		// Possible Mirror States to be used as desiredStates:
+		// mirrored, broken or uninitialized
+		if !utils.SliceContainsValue(validStates, strings.ToLower(desiredState)) {
+			return nil, "", fmt.Errorf("invalid desired mirror state was passed to check mirror replication state (%s), possible values: (%+v)", desiredState, volumesreplication.PossibleValuesForMirrorState())
+		}
+
+		res, err := client.VolumesReplicationStatus(ctx, id)
+		if err != nil {
+			if !response.WasNotFound(res.HttpResponse) {
+				return nil, "", fmt.Errorf("retrieving replication status information from %s: %s", id, err)
+			}
+		}
+
+		// TODO: fix this refresh function to use strings instead of fake status codes
+		// Setting 200 as default response
+		response := 200
+		if res.Model != nil && res.Model.MirrorState != nil && strings.EqualFold(string(*res.Model.MirrorState), desiredState) {
+			// return 204 if state matches desired state
+			response = 204
+		}
+
+		return res, strconv.Itoa(response), nil
+	}
+}
+
+func netappVolumeReplicationStateRefreshFunc(ctx context.Context, client *volumesreplication.VolumesReplicationClient, id volumesreplication.VolumeId) pluginsdk.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		res, err := client.VolumesReplicationStatus(ctx, id)
+		if err != nil {
+			if response.WasBadRequest(res.HttpResponse) && (strings.Contains(strings.ToLower(err.Error()), "deleting") || strings.Contains(strings.ToLower(err.Error()), "volume replication missing or deleted")) {
+				// This error can be ignored until a bug is fixed on RP side that it is returning 400 while the replication is in "Deleting" process
+				// TODO: remove this workaround when above bug is fixed
+			} else if !response.WasNotFound(res.HttpResponse) {
+				return nil, "", fmt.Errorf("retrieving replication status from %s: %s", id, err)
+			}
+		}
+		statusCode := "dropped connection"
+		if res.HttpResponse != nil {
+			statusCode = strconv.Itoa(res.HttpResponse.StatusCode)
+		}
+		return res, statusCode, nil
+	}
+}
+
+func translateTFSchedule(scheduleName string) string {
+	if strings.EqualFold(scheduleName, string(ReplicationSchedule10Minutes)) {
+		return string(volumegroups.ReplicationScheduleOneZerominutely)
+	}
+
+	return scheduleName
+}
+
+func translateSDKSchedule(scheduleName string) string {
+	if strings.EqualFold(scheduleName, string(volumegroups.ReplicationScheduleOneZerominutely)) {
+		return string(ReplicationSchedule10Minutes)
+	}
+
+	return scheduleName
+}
+
+func validateNetAppVolumeGroupVolumes(volumeList *[]volumegroups.VolumeGroupVolumeProperties, applicationType volumegroups.ApplicationType) []error {
```

Comment:
Ok, I'll also remove the applicationType then from the arguments since it will be dedicated to SAP-HANA application. When the time comes for other applications, we might have a common validation and make two calls, one for common stuff and one for specific application.